src/libraries/System.Text.RegularExpressions/src/dotnet-regular-expressions.md

```diff
@@ -0,0 +1,237 @@
+  * `protected abstract bool FindFirstChar()`
+  * `public Match? Scan(System.Text.RegularExpressions.Regex regex, string text...)` calls `FindFirstChar()` and `Go()`
+* Has a "quick" mode that does not instantiate any captures: used by `Regex.IsMatch(..)` which does not expose captures to the caller
+* Concrete instances created by `Match? Regex.Run(...)` calling either `RegexRunner CompiledRegexRunnerFactory.CreateInstance()` or newing up a `RegexInterpreter`
+
+### RegexInterpreter (is a RegexRunner)
+
+* See above. Note that this is sealed.
+
+### CompiledRegexRunner (is a RegexRunner)
+
+* See above.
+
+## Results
+
+### Match (public, is a Group)
+
+* Represents one match of the pattern: there may be several
+* Holds a `Regex` in order to call `NextMatch()`
+* Created by `RegexRunner`
+
+### Group (public, is a Capture)
+
+* Represents one capturing group from the match
+* Simple data holder
+
+### Capture (public)
+
+* Represents one of the potentially several captures from a capturing group; this is a .NET-only concept.
+* Simple data holder
+
+### MatchCollection (public)
+
+* Created by `Regex.Matches`
+* Lazily provides `Match`es
+
+### GroupCollection (public)
+
+* Created by `Match.Groups`
+* Lazily creates `Group`s
+
+### CaptureCollection (public)
+
+* Created by `Group.Captures`
+* Lazily creates `Capture`s
+
+### RegexParseException (is a ArgumentException)
+
+* Thrown when pattern is invalid
+* Contains `RegexParseError`
+
+### RegexMatchTimeoutException (public)
+
+* Thrown when timeout expires
+
+## Optimizations
+
+### Tree optimization
+
+* Every `RegexNode.AddChild()` calls `Reduce()` to attempt to optimize subtree as it is being assembled, and parsing ends with call to `RegexNode.FinalOptimize()` for some optimizations that require the entire tree. The goal is to make a functionally equivalent tree that can produce a more efficient program. With more detailed analysis of the tree and some creativity, more could be done here.
+
+### Testing character classes
+
+* Testing a character for membership of a character class can take a significant time in aggregate. Numerous optimizations have been made here. For example, originally it used a binary search, and now it attempts to use a bitmap where possible. More improvements here would likely be worthwhile.
+
+### Prefix matching
+
+* If the pattern begins with a literal, `FindFirstChar()` is used to run quickly to the next point in the text that matches that literal, without using the engine. If the literal is a single character, this can use `IndexOf()` which is vectorized; otherwise it uses `RegexBoyerMoore`. Future optimizations could, for example, handle an alternation of leading literals using the Aho–Corasick algorithm; or use `IndexOf` to find a low-probability char before matching the whole literal. These optimizations are likely to most help in the case of a large text, perhaps with few matches, and a pattern with leading large literals.
```

Comment:
> If the pattern begins with a literal

Not just a literal.  If it does begin with a literal string (a "multi"), then this is correct.  However, if it doesn't, the implementation will also try to find what characters could possibly start a match (e.g. if the expression is `abc|def`, then `a` and `d`), and it'll build a character class for that set.  FindFirstChar will then use that character class while finding the next potential starting point.